### PR TITLE
RDMA and SHARP plugin for NCCL

### PR DESCRIPTION
**Pull request by bureddy**

### CODE REVIEW

File: Makefile.am [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+#
+
+SUBDIRS = src
+
+EXTRA_DIST =
+EXTRA_DIST += autogen.sh
+EXTRA_DIST += include
+EXTRA_DIST += debian
+EXTRA_DIST += nccl-rdma-sharp-plugins.spec
+EXTRA_DIST += nccl-rdma-sharp-plugins.spec.in
```
Comment (mike-dubman) [inline Makefile.am]: why .in file in EXTRA? it should appear only in AC_OUTPUT as template and will be distributed automatically

File: configure.ac [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+AC_CHECK_LIB([ibverbs], [ibv_get_device_list],
+            [],[AC_MSG_FAILURE([libibverbs not found]);])
+
+
+# check for sharp
+AC_ARG_WITH([sharp],
+            AC_HELP_STRING([--with-sharp=PATH], [Path to non-standard SHARP installation]),
+             [CFLAGS="-I$withval/include $CFLAGS"
+              LDFLAGS="-L$withval/lib $LDFLAGS"],
+             [CFLAGS="-I/opt/mellanox/sharp/include $CFLAGS"
```
Comment (mike-dubman) [inline configure.ac]: it means - sharp from ofed
what about sharp from HPCX/sharp?
maybe if hpcx module loaded, you can detect HPCX_SHARPDIR var and use it and if not - go to mofed dir (if existst)

**Comment by mike-dubman** [comment]
does it mean that NCCL folks will remove verbs code from NCCL upsream?
Does it mean we take ownership on VERBS code as it is?
Does it mean we can replace VERBS code with UCX to ease support on rdma part?

File: Makefile.am (hunk 1 above)
Comment (bureddy) [reply Makefile.am]: fixed in new PR

File: configure.ac (hunk 2 above)
Comment (bureddy) [reply configure.ac]: fixed in new PR.

**Comment by bureddy** [comment]
Not decided on remove it from upstream yet. it will be there for next few releases for sure.
 we are complete owners of plug-in (VERBS + SHARP) and we can make UCX default with our plugin.

When we add UCX to plugin, i think it will add it add as a addition.  so user can have options to run in any of the following modes with plugin.
1.  VERBS only
2.  UCX only
3.  VERBS + SHARP
4.  UCX + SHARP

**Comment by mike-dubman** [comment]
Why need 2 copies of verbs plugin?
In upstream and here

**Comment by bureddy** [comment]
upstream verbs is not a plugin. it is directly integrated inside nccl core and it is part of libnccl.so
if NCCL runtime finds any network plugin in the path, it will have high priority .

**Comment by mike-dubman** [comment]
But why we need to keep same verbs code here and not call from upstream if needed?

**Comment by bureddy** [comment]
we can and that is VERBS only mode right?
you mean completely remove in the upstream? May be there can be some platforms (IWARP ?) where MOFED/HPCX is not installed. 
Anyway I think it is safe to overlap 1 or 2 releases to make decision on removing it in upstream if there is any real requirement. WDYT?

**Comment by mike-dubman** [comment]
But why we clone verbs code from upstream into plugin in the 1st place? Why not to keep it in upstream only?

**Comment by bureddy** [comment]
For sharp we need OOB channel for bootstrap.  the design was not clean by calling nccl internal APIs from external plugin for this OOB. So, we extracted pt2pt plugin (VERBS) also into Collective plug-in.  Now pt2pt plugin (VERBS) is OOB channel for Collective plug-in (SHARP).

**Comment by mike-dubman** [comment]
I see
Do they plan to expose oob api so sharp can call nccl core plugin for oob?
I

**Comment by bureddy** [comment]
AFAIK  there is no plan for this.  will check.

**Pull request merged**